roomserver/storage/postgres/user_room_keys_table.go

```diff
@@ -29,17 +31,33 @@ const userRoomKeysSchema = `
 CREATE TABLE IF NOT EXISTS roomserver_user_room_keys (     
     user_nid    INTEGER NOT NULL,
     room_nid    INTEGER NOT NULL,
-    pseudo_id_key BYTEA NOT NULL,
+    pseudo_id_key BYTEA NULL, -- may be null for users not local to the server
+    pseudo_id_pub_key BYTEA NOT NULL,
     CONSTRAINT roomserver_user_room_keys_pk PRIMARY KEY (user_nid, room_nid)
 );
 `
 
-const insertUserRoomKeySQL = `INSERT INTO roomserver_user_room_keys (user_nid, room_nid, pseudo_id_key) VALUES ($1, $2, $3)`
+const insertUserRoomPrivateKeySQL = `
+	INSERT INTO roomserver_user_room_keys (user_nid, room_nid, pseudo_id_key, pseudo_id_pub_key) VALUES ($1, $2, $3, $4)
+	ON CONFLICT ON CONSTRAINT roomserver_user_room_keys_pk DO UPDATE SET pseudo_id_key = roomserver_user_room_keys.pseudo_id_key
+	RETURNING (pseudo_id_key)
+`
+
+const insertUserRoomPublicKeySQL = `
+	INSERT INTO roomserver_user_room_keys (user_nid, room_nid, pseudo_id_pub_key) VALUES ($1, $2, $3)
+	ON CONFLICT ON CONSTRAINT roomserver_user_room_keys_pk DO UPDATE SET pseudo_id_pub_key = roomserver_user_room_keys.pseudo_id_pub_key
+	RETURNING (pseudo_id_pub_key)
+`
+
 const selectUserRoomKeySQL = `SELECT pseudo_id_key FROM roomserver_user_room_keys WHERE user_nid = $1 AND room_nid = $2`
 
+const selectUserNIDsSQL = `SELECT user_nid, pseudo_id_pub_key  FROM roomserver_user_room_keys WHERE pseudo_id_pub_key = ANY($1)`
```

Comment:
This isn't safe to do because it lacks a `WHERE room_nid` clause. The protocol does not say that pseudo ID keys need to be globally unique.